Mesh Divisions system to split the mesh in arbitrary ways

GiudGiud opened the pull request:
refs https://github.com/idaholab/moose/issues/25901

I'm delivering this with quite a few items to hope to see some adoption. I think it worked for Times & Positions to do that.
Realistically while this makes for a long review it also makes it efficient because the logic is similar everywhere

I am not including the hex lattice (which is the one I need) because there's a need for a utility and we should do this right once and for all. I can make do with nearest-positions for the transfer most likely.

GiudGiud committed:
Add a base class for mesh divisions, refs #25901
- public getters for which division and number of divisions
- meshChanged causes new initialize

GiudGiud committed:
Add Mesh Divisions on the problem for now

GiudGiud committed:
Add machinery to create MeshDivisions: new action

GiudGiud committed:
Add an auxkernel to visualize mesh divisions

GiudGiud committed:
Add block meshDivision refs #25901

GiudGiud committed:
Add mesh division by extra ID
A single type of extra ID is chosen, and the sorted values are used as bins

GiudGiud committed:
Add more APIs on positions to prepare nearest positions division

GiudGiud committed:
Add mesh division by nearest neighbor region from Positions

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1871589/) on e62b7fe wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25998/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25998/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f07619acc231d07003f7625ad27acd66306f4c62`

GiudGiud committed:
Add mesh division based on a Cartesian lattice

GiudGiud committed:
Add mesh division based on uniformly binned values of a functor

GiudGiud committed:
Add doc and test for subdomain-based division

GiudGiud committed:
Add doc and test for extra element id based division

GiudGiud committed:
Add doc and test for cartsian grid based division

GiudGiud committed:
Add a syntax page for mesh divisions

GiudGiud committed:
Add doc and test for functor bins based division

GiudGiud committed:
Add doc and test for nearest positions based division

GiudGiud committed:
Add a spherical coordinate based mesh division

GiudGiud committed:
Add a cylindical coordinate based mesh division

GiudGiud committed:
Add doc and test for spherical coordinate mesh division

GiudGiud committed:
Add doc and test for cylindrical coordinate mesh division

GiudGiud committed:
Add an object to nest several divisions together: for example nearest position and cylindrical coordinates for pebbles

GiudGiud committed:
Add doc and test for a nested division mesh division

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1873952/) on 6081334 wanted to post the following:

The following commit(s) contain fixups:

- 6081334d9c: fixup!  Add doc and test for nearest positions based division

GiudGiud committed:
Only test in replicated mode as the sampling is not ordered the same with distributed meshes

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1874765/) on 641a6db wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25998/site)

This comment will be updated on new commits.

GiudGiud committed:
Address test failures:
- initialize pointer
- modify bad test that ends up enforcing the list of systems
- add doco for action

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1879048/) on 594824a wanted to post the following:

The following commit(s) contain fixups:

- 594824a4df: fixup!  Add doc and test for nearest positions based division

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1879084/) on 594824a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>2f9664</th>
      <th colspan=4>#25998 594824</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.69%</td>
      <td>85.65%</td>
      <td>-0.04%</td>
      <td>76.60%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92782</td>
      <td>93159</td>
      <td>+377</td>
      <td>383</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15490</td>
      <td>15607</td>
      <td>+117</td>
      <td>117</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25998/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25998/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25998/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 76.60% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1879258/) on 6bc9532 wanted to post the following:

The following commit(s) contain fixups:

- 6bc95320d7: fixup!  Add doc and test for nearest positions based division

GiudGiud commented:
Threaded failure is real and I'm thinking about it but otherwise this can be looked at

GiudGiud committed:
Add a PP to count the number of mesh divisions

GiudGiud committed:
Make mesh divisions be replicated per thread:
- there is no guarantee that the objects calling mesh divsions are NON-threaded
- if the callers are threaded, any object called by the mesh division should be thread-safe, no guarantee about that either
unless we have a copy per thread

Address Alex's review

- add missing class description
- more constness
- explicit virtual when overriding
- use set_union instead of copying into a vector

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
I can add more testing (with a new potentially useful object) once https://github.com/idaholab/moose/pull/25872 is in

lindsayad commented on framework/src/meshdivisions/NearestPositionsDivision.C:
```diff
@@ -0,0 +1,55 @@
+  params.addClassDescription("Divide the mesh by ");
+  params.addRequiredParam<PositionsName>(
+      "positions",
+      "The name of the Positions object to form the nearest-neighbor partition/division with");
```
```suggestion
      "The name of the Positions object to form the nearest-neighbor division with");
```

lindsayad commented on framework/src/meshdivisions/NearestPositionsDivision.C:
```diff
@@ -0,0 +1,55 @@
+unsigned int
+NearestPositionsDivision::divisionIndex(const Elem & elem) const
+{
+  bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```
```suggestion
  const bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```

lindsayad commented on framework/src/meshdivisions/NearestPositionsDivision.C:
```diff
@@ -0,0 +1,55 @@
+unsigned int
+NearestPositionsDivision::divisionIndex(const Point & pt) const
+{
+  bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```
```suggestion
  const bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```

lindsayad commented on framework/src/meshdivisions/NestedDivision.C:
```diff
@@ -0,0 +1,78 @@
+    tot_divs *= _num_divs[i];
+    if (_num_divs[i] == 0)
+      mooseError("Nested division",
+                 getParam<std::vector<MeshDivisionName>>("divisions")[i],
```
```suggestion
                 _divisions[i]->name(),
```

lindsayad commented on framework/src/meshdivisions/NestedDivision.C:
```diff
@@ -0,0 +1,78 @@
+    if (_num_divs[i] == 0)
+      mooseError("Nested division",
+                 getParam<std::vector<MeshDivisionName>>("divisions")[i],
+                 "now has zero bins");
```
```suggestion
                 "' has zero bins");
```

lindsayad commented on framework/src/meshdivisions/NestedDivision.C:
```diff
@@ -0,0 +1,78 @@
+    _num_divs[i] = _divisions[i]->getNumDivisions();
+    tot_divs *= _num_divs[i];
+    if (_num_divs[i] == 0)
+      mooseError("Nested division",
```
```suggestion
      mooseError("Nested division '",
```

lindsayad commented on framework/src/meshdivisions/NestedDivision.C:
```diff
@@ -0,0 +1,78 @@
+{
+  unsigned int index = 0;
+  unsigned int running_product = 1;
+  auto N_divs = _divisions.size();
```
```suggestion
  const auto N_divs = _divisions.size();
```

lindsayad commented on framework/src/meshdivisions/NestedDivision.C:
```diff
@@ -0,0 +1,78 @@
+{
+  unsigned int index = 0;
+  unsigned int running_product = 1;
+  auto N_divs = _divisions.size();
```
```suggestion
  const auto N_divs = _divisions.size();
```

lindsayad commented on framework/src/meshdivisions/SphericalGridDivision.C:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  const auto not_found = MooseMeshDivision::INVALID_DIVISION_INDEX;
+  unsigned int ir = not_found;
```
```suggestion
  auto ir = not_found;
```

lindsayad commented on framework/src/postprocessors/NumMeshDivisions.C:
```diff
@@ -0,0 +1,37 @@
+
+// MOOSE includes
+#include "NumMeshDivisions.h"
+#include "MeshDivision.h"
```
```suggestion
#include "MeshDivision.h"
#include "FEProblemBase.h"
```

lindsayad commented on test/tests/meshdivisions/tests:
```diff
@@ -0,0 +1,69 @@
+      type = CSVDiff
+      input = 'nested_division.i'
+      csvdiff = 'nested_division_out_div_out_0001.csv'
+      detail = 'a nesting of mesh divisions.'
```
```suggestion
      detail = 'and a nesting of mesh divisions.'
```

lindsayad commented on unit/src/MooseServerTest.C:
```diff
@@ -933,7 +933,7 @@ TEST_F(MooseServerTest, CompletionDocumentRootLevel)
   EXPECT_EQ(request_id, response_id);
 
-  EXPECT_EQ(46u, completions_array.size());
+  EXPECT_EQ(47u, completions_array.size());
```
I agree with your issue that this is unfortunate

lindsayad commented on framework/doc/content/source/meshdivisions/NearestPositionsDivision.md:
```diff
@@ -0,0 +1,18 @@
+object specified in the [!param](/MeshDivisions/NearestPositionsDivision/positions) parameter.
+
+!alert note
+For points laying within the standard tolerance of an internal boundary of the bins,
```
```suggestion
For points lying within the standard tolerance of an internal boundary of the bins,
```

lindsayad commented on framework/doc/content/source/meshdivisions/SphericalGridDivision.md:
```diff
@@ -0,0 +1,25 @@
+The number of mesh divisions/regions is the product of the number of bins in each coordinate, which currently only
+includes the radial direction.
+
+Points that lay outside the spherical grid may be assigned to the closest grid outer bins (on the inner or outer shell)
```
```suggestion
Points that lie outside the spherical grid may be assigned to the closest grid outer bins (on the inner or outer shell)
```

lindsayad commented on framework/doc/content/source/meshdivisions/SphericalGridDivision.md:
```diff
@@ -0,0 +1,25 @@
+parameter.
+
+!alert note
+We have not implemented neither binning in the azimuthal and toroidal coordinates nor restrictions in those angular coordinates.
```
```suggestion
We have not implemented binning in the azimuthal nor toroidal coordinates nor restrictions in those angular coordinates.
```

lindsayad commented on framework/doc/content/source/meshdivisions/SphericalGridDivision.md:
```diff
@@ -0,0 +1,25 @@
+This is a desirable extension of this object.
+
+!alert note
+For points laying within the standard tolerance of an internal boundary of the spherical grid, this object
```
```suggestion
For points lying within the standard tolerance of an internal boundary of the spherical grid, this object
```

lindsayad reviewed (changes requested):
looks great!

GiudGiud commented on framework/include/meshdivisions/FunctorBinnedValuesDivision.h:
```diff
@@ -0,0 +1,47 @@
+   */
+  unsigned int getBinIndex(Real value, const Point & pt) const;
+
+protected:
```
```suggestion
```

GiudGiud commented on framework/include/meshdivisions/FunctorBinnedValuesDivision.h:
```diff
@@ -0,0 +1,47 @@
+   * @param value functor value to bin
+   * @param pt point where the functor is being evaluated, useful for warnings
+   */
+  unsigned int getBinIndex(Real value, const Point & pt) const;
```
no. moving it

GiudGiud commented on framework/include/meshdivisions/FunctorBinnedValuesDivision.h:
```diff
@@ -0,0 +1,47 @@
+  void initialize() override;
+  unsigned int divisionIndex(const Point & pt) const override;
+  unsigned int divisionIndex(const Elem & elem) const override;
+
```
```suggestion
protected:

```

GiudGiud commented on framework/include/meshdivisions/MeshDivision.h:
```diff
@@ -0,0 +1,78 @@
+#include "Restartable.h"
+#include "MeshChangedInterface.h"
+
+// libMesh
```
```suggestion
```

GiudGiud commented on framework/src/auxkernels/MeshDivisionAux.C:
```diff
@@ -0,0 +1,45 @@
+    _mesh_division(_c_fe_problem.getMeshDivision(getParam<MeshDivisionName>("mesh_division"), _tid))
+{
+  // Check family types for reasonable choices
+  if (isNodal() && _var.feType().order != FIRST)
+    paramError("Variable order " + Moose::stringify(_var.feType().order) +
+               " unsupported for nodal variable");
```
```suggestion
```

GiudGiud commented on framework/src/auxkernels/MeshDivisionAux.C:
```diff
@@ -0,0 +1,45 @@
+  if (isNodal() && _var.feType().order != FIRST)
+    paramError("Variable order " + Moose::stringify(_var.feType().order) +
+               " unsupported for nodal variable");
+  else if (_var.feType().order != CONSTANT)
```
```suggestion
  if (!_var.isNodal() && _var.feType().order != CONSTANT)
```

GiudGiud commented on framework/src/meshdivisions/NearestPositionsDivision.C:
```diff
@@ -0,0 +1,55 @@
+NearestPositionsDivision::validParams()
+{
+  InputParameters params = MeshDivision::validParams();
+  params.addClassDescription("Divide the mesh by ");
```
```suggestion
  params.addClassDescription("Divide the mesh using a nearest-point / voronoi algorithm, with the points coming from a Positions object");
```

GiudGiud committed:
Avoid renumbering to be able to test mesh divisions in parallel-distributed
